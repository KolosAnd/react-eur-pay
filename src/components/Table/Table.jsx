import React, {useEffect, useState} from "react";
import './table.style.scss';
import {Fees} from "../Fees/Fees.jsx";
import {DownloadButton} from "../DownloadButton/DownloadButton.jsx";

export const Table = () => {
    const [operations, setOperations] = useState([]);
    const [feesArray, setFeesArray] = useState([]);

    const [error, setError] = useState(false);

    useEffect(() => {
        fetch(process.env.PUBLIC_URL + '/input.json')
            .then(response => response.json())
            .then(data => setOperations(data.operations))
            .catch(error => console.error('Error loading the data', error));
    }, []);

    return (
        <>
            {!error ?
                <div className="table-wrap">
                    <div className="tables-container">
                        <table border="1" className="table">
                            <thead>
                            <tr>
                                <th>Date</th>
                                <th>User ID</th>
                                <th>User Type</th>
                                <th>Type</th>
                                <th>Amount</th>
                            </tr>
                            </thead>
                            <tbody>
                            {operations.map((item, index) => (
                                <tr key={index}>
                                    <td>{item.date}</td>
                                    <td>{item.user_id}</td>
                                    <td>{item.user_type}</td>
                                    <td>{item.type}</td>
                                    <td>{item.operation.amount.toFixed(2)}</td>
                                </tr>
                            ))}
                            </tbody>
                        </table>

                            <Fees operations={operations} setFees={setFeesArray} setError={setError} />

                    </div>
                    {feesArray?.length > 0 &&
                        <DownloadButton feesArray={feesArray} />
                    }
                </div>
                :
                <div className="error-wrap">
                    <h2>
                        Something wrong with api
                    </h2>
                </div>
            }
        </>

    )
}
