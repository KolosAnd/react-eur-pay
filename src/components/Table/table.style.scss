

.table-wrap{
  width: 100%;

  .tables-container{
    display: flex;
    width: fit-content;
    margin: 0 auto 20px;
  }
  td {
    padding: 4px 5px;
  }
  button {
    cursor: pointer;
    display: block;
    margin: 0 auto;
    background: rgba(255, 255, 255, 0.87);
    color: #242424;
    padding: 4px 12px;
  }
}

.error-wrap {
  h2 {
    text-align: center;
  }
}


