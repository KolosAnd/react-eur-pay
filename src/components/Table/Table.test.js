import { render, screen, waitFor } from '@testing-library/react';
import { Table } from './Table';
import renderer from 'react-test-renderer';
import {QueryClient, QueryClientProvider,} from "@tanstack/react-query";
import fetchMock from 'jest-fetch-mock';

fetchMock.enableMocks();

beforeEach(() => {
    fetch.resetMocks();
});
const queryClient = new QueryClient()

test('Table default', () => {
    const component = renderer.create(
        <QueryClientProvider client={queryClient}>
            <Table/>
        </QueryClientProvider>
        ,);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});


describe('Table component', () => {
    it('loads and displays the operations data correctly', async () => {
        fetchMock.mockResponseOnce(JSON.stringify({
            operations: [
                { date: '2020-01-01', user_id: 1, user_type: 'natural', type: 'cash_out', operation: { amount: 100.00 } }
            ]
        }));

        render(
        <QueryClientProvider client={queryClient}>
            <Table/>
        </QueryClientProvider>);

    });

    it('renders DownloadButton when feesArray has items', async () => {
        fetchMock.mockResponseOnce(JSON.stringify({
            operations: [
                { date: '2020-01-01', user_id: 1, user_type: 'natural', type: 'cash_out', operation: { amount: 100.00 } }
            ]
        }));

        render(
        <QueryClientProvider client={queryClient}>
            <Table/>
        </QueryClientProvider>);

        await waitFor(() => {
            expect(screen.queryByRole('button', { name: /download output\.json/i })).not.toBeInTheDocument();
        });
    });
});
