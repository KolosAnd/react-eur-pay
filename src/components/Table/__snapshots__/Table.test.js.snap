// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table default 1`] = `
<div
  className="table-wrap"
>
  <div
    className="tables-container"
  >
    <table
      border="1"
      className="table"
    >
      <thead>
        <tr>
          <th>
            Date
          </th>
          <th>
            User ID
          </th>
          <th>
            User Type
          </th>
          <th>
            Type
          </th>
          <th>
            Amount
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <table
      border="1"
      className="table"
    >
      <thead>
        <tr>
          <th>
            Fees
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
